File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
from collections import deque

def solution(name):
    n = len(name)
    answer = 0
    flag = 0
    
    # 각 알파벳을 A로 만들기
    for idx, ch in enumerate(name):
        if ch == 'A': continue
        answer += min(
            ord(ch) - ord('A'),
            ord('Z') + 1 - ord(ch)
        )
        flag |= (1 << idx)
    
    # 현재 위치, 이동 횟수, 방문 필요 여부
    deq = deque([[0, 0, flag]])
    
    # BFS로 좌우 이동 횟수를 구함
    while deq:
        pos, cnt, nextFlag = deq.popleft()
        nextFlag &= ~(1 << pos)
        if nextFlag == 0: return answer + cnt
        deq.append([(pos + 1) % n, cnt + 1, nextFlag])
        deq.append([(pos - 1 + n) % n, cnt + 1, nextFlag])

File: 프로그래머스/2/42860. 조이스틱/조이스틱.cpp
#include <string>
#include <vector>
#include <queue>
#include <tuple>

using namespace std;
using tp = tuple<int, int, int>; // cnt, pos, targets

int solution(string name) {
    size_t n = name.size();
    
    int cnt = 0;
    int targets = 0;
    
    for (int i = 0; i < n; i++) {
        char ch = name[i];
        
        int up = ch - 'A';
        int down = 'Z' + 1 - ch;
        int dist = min(up, down);
        
        if (!dist) continue;
        
        cnt += dist;
        targets |= (1 << i);
    }
    
    queue<tp> q;
    q.emplace(0, 0, targets);
    
    while (!q.empty()) {
        auto [_cnt, pos, _targets] = q.front();
        q.pop();
        
        _targets -= (_targets & (1 << pos));
        if (!_targets) return cnt + _cnt;
        
        int left = pos - 1;
        if (left == -1) left = n - 1;
        q.emplace(_cnt + 1, left, _targets);
        
        int right = pos + 1;
        if (right == n) right = 0;
        q.emplace(_cnt + 1, right, _targets);
    }
    
    return -1;
}